scipy/stats/_censored_data.py

```diff
@@ -0,0 +1,356 @@
+import numpy as np
+
+
+def _validate_1d(a, name):
+    if np.ndim(a) != 1:
+        raise ValueError(f'{name} must be a one-dimensional sequence')
+    if np.isnan(a).any():
+        raise ValueError(f'{name} must not contain nan.')
+    if np.isinf(a).any():
+        raise ValueError(f'{name} must contain only finite values')
+
+
+def _validate_intervals(intervals):
+    intervals = np.asarray(intervals)
+    if intervals.shape == (0,):
+        # The input was a sequence with length 0.
+        intervals = intervals.reshape((0, 2))
+    if intervals.ndim != 2 or intervals.shape[-1] != 2:
+        raise ValueError('intervals must be a two-dimensional array'
+                         ' with shape (m, 2)')
+
+    if np.isnan(intervals).any():
+        raise ValueError('intervals must not contain nan.')
+    if np.isinf(intervals).all(axis=1).any():
+        raise ValueError('In each row in intervals, both values must not'
+                         ' be infinite.')
+    if (intervals[:, 0] > intervals[:, 1]).any():
+        raise ValueError('In each row of intervals, the left value must not'
+                         ' exceed the right value')
+
+    uncensored = intervals[:, 0] == intervals[:, 1]
+    left_censored = np.isinf(intervals[:, 0])
+    right_censored = np.isinf(intervals[:, 1])
+    interval_censored = np.isfinite(intervals).all(axis=1) & ~uncensored
+
+    x2 = intervals[uncensored, 0]
+    left2 = intervals[left_censored, 1]
+    right2 = intervals[right_censored, 0]
+    intervals2 = intervals[interval_censored]
+
+    return x2, left2, right2, intervals2
+
+
+def _validate_x_censored(x, censored):
+    x = np.asarray(x)
+    if x.ndim != 1:
+        raise ValueError('x must be one-dimensional.')
+    censored = np.asarray(censored)
+    if censored.ndim != 1:
+        raise ValueError('censored must be one-dimensional')
```

Comment:
I'd prefer that we just broadcast `x` and `censored`, but if I get really annoyed, I'll change it later.